// _egg-type-graphql@1.8.1@egg-type-graphql/lib/addDirective.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const graphql_1 = require("graphql");
const graphqlLanguage = tslib_1.__importStar(require("graphql/language"));
const graphqlType = tslib_1.__importStar(require("graphql/type"));
const execution_1 = require("graphql/execution");
const DirectiveLocation = graphqlLanguage.DirectiveLocation || graphqlType.__DirectiveLocation;
const BUILT_IN_DIRECTIVES = ['deprecated', 'skip', 'include'];
function forEachField(schema, fn) {
    const typeMap = schema.getTypeMap();
    Object.keys(typeMap).forEach(typeName => {
        const type = typeMap[typeName];
        // TODO: maybe have an option to include these?
        if (!graphql_1.getNamedType(type).name.startsWith('__') &&
            type instanceof graphql_1.GraphQLObjectType) {
            const fields = type.getFields();
            Object.keys(fields).forEach(fieldName => {
                const field = fields[fieldName];
                fn(field, typeName, fieldName);
            });
        }
    });
}
function getFieldResolver(field) {
    const resolver = field.resolve || graphql_1.defaultFieldResolver;
    return resolver.bind(field);
}
function createAsyncResolver(field) {
    const originalResolver = getFieldResolver(field);
    const resolverFn = async (source, args, context, info) => originalResolver(source, args, context, info);
    return resolverFn;
}
function getDirectiveInfo(directive, resolverMap, schema, location, variables) {
    const name = directive.name.value;
    const Directive = schema.getDirective(name);
    if (typeof Directive === 'undefined') {
        throw new Error(`Directive @${name} is undefined. ` +
            'Please define in schema before using.');
    }
    if (!Directive || !Directive.locations.includes(location)) {
        throw new Error(`Directive @${name} is not marked to be used on "${location}" location. ` +
            `Please add "directive @${name} ON ${location}" in schema.`);
    }
    const resolver = resolverMap[name];
    if (!resolver && !BUILT_IN_DIRECTIVES.includes(name)) {
        throw new Error(`Directive @${name} has no resolver.` +
            'Please define one using createFieldExecutionResolver().');
    }
    const args = execution_1.getDirectiveValues(Directive, { directives: [directive] }, variables) || {};
    return { args, resolver };
}
function createFieldExecutionResolver(field, resolverMap, schema) {
    if (!field.astNode)
        return getFieldResolver(field);
    const { directives } = field.astNode;
    if (!directives || !directives.length)
        return getFieldResolver(field);
    return directives.reduce((prevResolver, directive) => {
        const directiveInfo = getDirectiveInfo(directive, resolverMap, schema, DirectiveLocation.FIELD_DEFINITION);
        const resolverFn = (source, args, context, info) => directiveInfo.resolver({
            resolve: () => prevResolver(source, args, context, info),
            source,
            args: directiveInfo.args,
            context,
            info,
        });
        return resolverFn;
    }, createAsyncResolver(field));
}
function createFieldResolver(field, resolverMap, schema) {
    const originalResolver = getFieldResolver(field);
    const asyncResolver = createAsyncResolver(field);
    const resolverFn = (source, args, context, info) => {
        const { directives } = info.fieldNodes[0];
        if (!directives || !directives.length) {
            return originalResolver(source, args, context, info);
        }
        const fieldResolver = directives.reduce((prevResolver, directive) => {
            const directiveInfo = getDirectiveInfo(directive, resolverMap, schema, DirectiveLocation.FIELD, info.variableValues);
            const value = () => directiveInfo.resolver({
                resolve: () => prevResolver(source, args, context, info),
                source,
                args: directiveInfo.args,
                context,
                info,
            });
            return () => value();
        }, asyncResolver);
        return fieldResolver(source, args, context, info);
    };
    return resolverFn;
}
function addDirectiveResolveFunctionsToSchema(schema, resolverMap) {
    if (typeof resolverMap !== 'object') {
        throw new Error(`Expected resolverMap to be of type object, got ${typeof resolverMap}`);
    }
    if (Array.isArray(resolverMap)) {
        throw new Error('Expected resolverMap to be of type object, got Array');
    }
    forEachField(schema, field => {
        field.resolve = createFieldExecutionResolver(field, resolverMap, schema);
        field.resolve = createFieldResolver(field, resolverMap, schema);
    });
    return schema;
}
exports.default = addDirectiveResolveFunctionsToSchema;
