// _egg-type-graphql@1.8.1@egg-type-graphql/lib/GraphQLServer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = require("path");
const fs_1 = require("fs");
const fs_jetpack_1 = require("fs-jetpack");
const typedi_1 = require("typedi");
const apollo_server_koa_1 = require("apollo-server-koa");
const type_graphql_1 = require("type-graphql");
const graphql_1 = require("graphql");
const isodate_1 = require("./scalars/isodate");
const timestamp_1 = require("./scalars/timestamp");
const addDirective_1 = tslib_1.__importDefault(require("./addDirective"));
function isClass(func) {
    return typeof func === 'function' && /^class\s/.test(Function.prototype.toString.call(func));
}
class GraphQLServer {
    constructor(app, server) {
        this.app = app;
        this.koaServer = server;
        this.graphqlConfig = app.config.typeGraphQL;
    }
    getDirectives() {
        const { baseDir } = this.app;
        const directivesDir = path_1.join(baseDir, 'app', 'directive');
        if (!fs_1.existsSync(directivesDir))
            return {};
        const matching = this.app.config.env === 'local' ? '*.ts' : '*.js';
        const files = fs_jetpack_1.find(directivesDir, { matching });
        return files.reduce((prev, cur) => {
            const directivePath = path_1.join(baseDir, cur);
            const name = path_1.basename(directivePath).replace(/(.ts)|(.js)$/, '');
            return Object.assign({}, prev, { [name]: require(directivePath).default });
        }, {});
    }
    getResolverClassFromFile(resolverPath) {
        const resolverModule = require(resolverPath);
        const resolvers = Object.keys(resolverModule).reduce((result, cur) => {
            if (isClass(resolverModule[cur]) && resolverModule[cur].toString().includes('Resolver')) {
                if (!result.includes(resolverModule[cur])) {
                    result.push(resolverModule[cur]);
                }
            }
            return result;
        }, []);
        return resolvers;
    }
    loadResolvers() {
        const { baseDir } = this.app;
        const appDir = path_1.join(baseDir, 'app');
        let resolvers = [];
        if (!fs_1.existsSync(appDir)) {
            this.app.logger.warn('[egg-type-graphql]', '缺少 resolver 文件');
            return [];
        }
        // TODO: handle other env
        const matching = this.app.config.env === 'local' ? '*.resolver.ts' : '*.resolver.js';
        const files = fs_jetpack_1.find(appDir, { matching });
        if (!files.length) {
            this.app.logger.error('[egg-type-graphql]', '缺少 resolver');
            return [];
        }
        try {
            for (const file of files) {
                const resolverPath = path_1.join(baseDir, file);
                const resolversFromFile = this.getResolverClassFromFile(resolverPath);
                if (!resolversFromFile.length) {
                    this.app.logger.error('[egg-type-graphql]', `${file} 文件必须存在至少一Resolver`);
                }
                resolvers = [...resolvers, ...resolversFromFile];
            }
        }
        catch (e) {
            this.app.logger.error('[egg-type-graphql]', e);
        }
        return resolvers;
    }
    setDirective(schema) {
        const { typeDefs } = this.graphqlConfig;
        if (typeDefs) {
            schema = graphql_1.extendSchema(schema, graphql_1.parse(typeDefs));
        }
        const resolverMap = this.getDirectives();
        if (Object.keys(resolverMap).length) {
            return addDirective_1.default(schema, resolverMap || {});
        }
        return schema;
    }
    async getSchema() {
        const { scalarsMap = [], dateScalarMode } = this.graphqlConfig;
        let schema;
        const resolvers = this.loadResolvers();
        if (!resolvers.length)
            return null;
        const defaultScalarMap = [
            {
                type: Date,
                scalar: dateScalarMode === 'timestamp' ? timestamp_1.GraphQLTimestamp : isodate_1.GraphQLISODateTime,
            },
        ];
        try {
            schema = await type_graphql_1.buildSchema({
                resolvers,
                dateScalarMode: 'isoDate',
                scalarsMap: [...defaultScalarMap, ...scalarsMap],
                emitSchemaFile: true,
                validate: this.graphqlConfig.validate,
                globalMiddlewares: this.graphqlConfig.globalMiddlewares || [],
                container: typedi_1.Container,
            });
            return this.setDirective(schema);
        }
        catch (e) {
            this.app.logger.error('[egg-type-graphql]', e);
        }
    }
    async start() {
        const schema = await this.getSchema();
        if (!schema)
            return null;
        const apolloConfig = {
            schema,
            tracing: false,
            context: ({ ctx }) => ctx,
            playground: {
                settings: {
                    'request.credentials': 'include',
                },
            },
            introspection: true,
        };
        const server = new apollo_server_koa_1.ApolloServer(apolloConfig);
        server.applyMiddleware({
            app: this.app,
            path: this.graphqlConfig.router,
            cors: false,
        });
        server.installSubscriptionHandlers(this.koaServer);
        this.app.logger.info('[egg-type-graphql] GraphQL server started');
    }
}
exports.default = GraphQLServer;
